nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceNode.java

```diff
@@ -229,7 +244,106 @@ public ControllerServiceState getState() {
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Will atomically disable this service by invoking its @OnDisabled operation.
+     * It uses CAS operation on {@link #stateRef} to transition this service
+     * from ENABLED to DISABLING state. If such transition succeeds the service
+     * will be de-activated (see {@link ControllerServiceNode#isActive()}).
+     * If such transition doesn't succeed (the service is still in ENABLING state)
+     * then the service will still be transitioned to DISABLING state to ensure that
+     * no other transition could happen on this service. However in such event
+     * (e.g., its @OnEnabled finally succeeded), the {@link #enable(ScheduledExecutorService, long, Heartbeater)}
+     * operation will initiate service disabling javadoc for (see {@link #enable(ScheduledExecutorService, long, Heartbeater)}
+     * <br>
+     * Upon successful invocation of @OnDisabled this service will be transitioned to
+     * DISABLED state.
+     */
+    @Override
+    public void disable(final ScheduledExecutorService scheduler, final Heartbeater heartbeater) {
+        this.active.set(false); // de-activating regardless of CAS operation
+                                // that follows since this operation will always result in service state being DISABLING
+        if (this.stateRef.compareAndSet(ControllerServiceState.ENABLED, ControllerServiceState.DISABLING)) {
+            final ConfigurationContext configContext = new StandardConfigurationContext(this, this.serviceProvider, null);
+            scheduler.execute(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        ReflectionUtils.invokeMethodsWithAnnotation(OnDisabled.class, StandardControllerServiceNode.this.getControllerServiceImplementation(), configContext);
+                    } catch (Exception e) {
+                        final Throwable cause = e instanceof InvocationTargetException ? e.getCause() : e;
+                        final ComponentLog componentLog = new SimpleProcessLogger(getIdentifier(), StandardControllerServiceNode.this);
+                        componentLog.error("Failed to invoke @OnDisabled method due to {}", cause);
+                        LOG.error("Failed to invoke @OnDisabled method of {} due to {}", getControllerServiceImplementation(), cause.toString());
+                        ReflectionUtils.quietlyInvokeMethodsWithAnnotation(OnDisabled.class, getControllerServiceImplementation(), configContext);
+                    } finally {
+                        stateRef.set(ControllerServiceState.DISABLED);
+                        heartbeater.heartbeat();
+                    }
+                }
+            });
+        } else {
+            this.stateRef.compareAndSet(ControllerServiceState.ENABLING, ControllerServiceState.DISABLING);
```

Comment:
@olegz I think there was a case that was missed here. Assume that we have two threads: E at line 280 and D and line 326. In this case, if D executes it will return false from compareAndSet. Then, if thread E executes, it will change the state from ENABLING to ENABLED. Thread D then executes line 346 and gets a value of false. At this point, Thread D returns from the disable() method, but it never did disable the service. This appears to be a race condition that we need to tackle.